### PR TITLE
Update our tools to run on >NET 8

### PR DESCRIPTION
**Pull request by BillWagner**
Changes in order of commits:

1. Update the runtime to .NET 8. Enable nullable warnings and implicit usings in all projects.
2. Update all NuGet packages to the latest released version (See note below).
3. Remove unnecessary usings and convert namespace declarations to file scoped namespaces.
4. Add an editorconfig for us to develop our own standard.
5. Update Yaml files for workflows to install the .NET 8 SDK.

Note on NuGet packages:

I tried updating the FSharp.Formatting package. The Markdown converter is using version 4.0.1, and the latest is on version 19. Version 20 is in alpha.  There are a small number of source breaking changes in APIs. Those were straightforward. You can see them reverted in the 2nd commit. The reason I reverted them is that the behavior is different, and I couldn't get everything working in a reasonable amount of time. There's two behavior issues stopping the update:

1. Instead of putting the space at the of a text span before *end note*, the updated library truncates the text node and adds a second sibling text node with a single space. This is similar for a number of formatted nodes (examples, etc).
2. When the first column of a table contains code bold code text, the converter renders a text node with `**` in the word doc.

I'll make an issue for these so we can work on that when either Jon or I have time.

### CODE REVIEW

File: tools/ExampleTester/ExampleTester.csproj [hunk 1]
```diff
@@ -2,15 +2,15 @@
...
     <ImplicitUsings>enable</ImplicitUsings>
     <Nullable>enable</Nullable>
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Build.Locator" Version="1.5.5" />
-    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="4.3.0" />
-    <PackageReference Include="Microsoft.CodeAnalysis.Workspaces.MSBuild" Version="4.3.0" />
+    <PackageReference Include="Microsoft.Build.Locator" Version="1.6.10" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="4.7.0" />
```
Comment (jskeet) [inline tools/ExampleTester/ExampleTester.csproj]: These are now at 4.8.0 (as I've just been updating our libraries today...) - it's just been reelased.
Might as well go to the latest while we're here.

File: tools/MarkdownConverter.Tests/MarkdownConverter.Tests.csproj [hunk 2]
```diff
@@ -1,8 +1,9 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>net6.0</TargetFramework>
+    <TargetFramework>net8.0</TargetFramework>
     <Nullable>enable</Nullable>
+	<ImplicitUsings>enable</ImplicitUsings>
```
Comment (jskeet) [inline tools/MarkdownConverter.Tests/MarkdownConverter.Tests.csproj]: Nit: outdent

File: tools/MarkdownConverter.Tests/MarkdownConverter.Tests.csproj [hunk 3]
```diff
@@ -13,14 +14,14 @@
...
-    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.3">
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />
+    <PackageReference Include="XMLUnit.Core" Version="2.9.2" />
+    <PackageReference Include="xunit" Version="2.6.2" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.5.4">
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
       <PrivateAssets>all</PrivateAssets>
     </PackageReference>
-    <PackageReference Include="coverlet.collector" Version="3.1.0">
+    <PackageReference Include="coverlet.collector" Version="6.0.0">
```
Comment (jskeet) [inline tools/MarkdownConverter.Tests/MarkdownConverter.Tests.csproj]: Should we just get rid of this? It's not like we're actually going to use coverage on this...

File: tools/MarkdownConverter/MarkdownConverter.csproj [hunk 4]
```diff
@@ -2,17 +2,18 @@
...
   <ItemGroup>
     <PackageReference Include="csharp2colorized" Version="1.0.3" />
-    <PackageReference Include="DocumentFormat.OpenXml" Version="2.15.0" />
-    <PackageReference Include="FSharp.Core" Version="6.0.1" />
-    <PackageReference Include="FSharp.Formatting" Version="4.1.0" />
-    <PackageReference Include="Microsoft.CodeAnalysis" Version="4.0.1" />
+    <PackageReference Include="DocumentFormat.OpenXml" Version="3.0.0" />
+    <PackageReference Include="FSharp.Core" Version="8.0.100" />
+    <PackageReference Include="FSharp.Formatting" Version="20.0.0-alpha-012" />
+    <PackageReference Include="Microsoft.CodeAnalysis" Version="4.7.0" />
```
Comment (jskeet) [inline tools/MarkdownConverter/MarkdownConverter.csproj]: 4.8.0 now.

**Review by jskeet — APPROVED** [review]
Just a few nits. Yes, we'll need to work out what to do about the F# Markdown package.

**Comment by jskeet** [comment]
I'm still getting warnings:

> warning NU1608: Detected package version outside of dependency constraint: FSharp.Formatting 19.1.1 requires FSharp.Core (= 7.0.400) but version FSharp.Core 8.0.101 was resolved.  
> warning NU1608: Detected package version outside of dependency constraint: FSharp.Compiler.Service 43.7.400 requires FSharp.Core (= 7.0.400) but version FSharp.Core 8.0.101 was resolved.

The commit for the update to Formatting v17 includes "Now, the warning on dependencies has gone away. (The F# formatter approves of the new F# compiler)" - have the warnings still gone away for you, even with 19.1.1?

**Comment by BillWagner** [comment]
> I'm still getting warnings:
> 
> > warning NU1608: Detected package version outside of dependency constraint: FSharp.Formatting 19.1.1 requires FSharp.Core (= 7.0.400) but version FSharp.Core 8.0.101 was resolved.
> > warning NU1608: Detected package version outside of dependency constraint: FSharp.Compiler.Service 43.7.400 requires FSharp.Core (= 7.0.400) but version FSharp.Core 8.0.101 was resolved.
> 
> The commit for the update to Formatting v17 includes "Now, the warning on dependencies has gone away. (The F# formatter approves of the new F# compiler)" - have the warnings still gone away for you, even with 19.1.1?

I am still getting it on the build, but not in the NuGet CLI. Interesting.

**Pull request merged**